feeder/feeder.go

```diff
@@ -0,0 +1,273 @@
+/*
+ * container-feeder: import Linux container images delivered as RPMs
+ * Copyright 2018 SUSE LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package feeder
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	wlk "github.com/kubic-project/container-feeder/walker"
+	log "github.com/sirupsen/logrus"
+)
+
+// the default path of container-feeder.json config
+var configFile = "/etc/container-feeder.json"
+
+// special type to load the container-feeder.json config
+type FeederConfig struct {
+	Target string `json:"feeder-target,omitempty"`
+}
+
+// loadConfig loads and returns the container-feeder.json config
+func loadConfig() (FeederConfig, error) {
+	config := FeederConfig{}
+
+	file, err := ioutil.ReadFile(configFile)
+	if err != nil {
+		return config, err
+	}
+
+	if err := json.Unmarshal(file, &config); err != nil {
+		return config, err
+	}
+
+	return config, nil
+}
+
+type FailedImportError struct {
+	Image string
+	Error error
+}
+
+type FeederLoadResponse struct {
+	SuccessfulImports []string
+	FailedImports     []FailedImportError
+}
+
+/* Image metadata type JSON schema:
+{
+  "image": {
+    "name": "",
+    "tags": [ ],
+    "file": ""
+  }
+}
+For example:
+{
+  "image": {
+    "name": "opensuse/salt-api",
+    "tags": [ "13", "13.0.1", "latest" ],
+    "file": "salt-api-2017.03-docker-images.x86_64.tar.xz"
+  }
+}
+*/
+// MetadataType struct to handle JSON schema
+type MetadataType struct {
+	Image ImageType
+}
+
+// ImageType struct to handle JSON schema
+type ImageType struct {
+	Name string   `json:"name"`
+	Tags []string `json:"tags"`
+	File string   `json:"file"`
+}
+
+// Feeder is a generalized interface that Container Feeders must implement
+type Feeder interface {
+	Images() ([]string, error)
+	LoadImage(string) (string, error)
+	TagImage(string, []string) error
+}
+
+// stringInSlice returns true if a is in list.
+func stringInSlice(a string, list []string) bool {
+	for _, b := range list {
+		if b == a {
+			return true
+		}
+	}
+	return false
+}
+
+// NewFeeder returns a new Container Feeder based on the specified type in
+// the container-feeder.json config (default: DockerFeeder)
+func NewFeeder() (Feeder, error) {
+	config, err := loadConfig()
+	if err != nil {
+		log.Errorf("Error loading config file, defaulting to docker feeder: %v", err)
+		config.Target = "docker"
+	}
+
+	if config.Target == "docker" || config.Target == "" {
```

Comment:
Maybe we could use `switch` statement here, in future this list can be longer
https://github.com/golang/go/wiki/Switch#switch

Reply:
Done